File: unity/GameGuest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using UnityEngine;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using NativeWebSocket;

// IMPORTANTE: Para WebGL necesitas instalar el paquete NativeWebSocket:
// En Unity, ve a Window > Package Manager > Add package from git URL...
// Ingresa: https://github.com/endel/NativeWebSocket.git#upm
public class GameGuest : MonoBehaviour
{
    [Header("Configuration")]
    public string wsUrl = "wss://tu-api-id.execute-api.region.amazonaws.com/stage";
    public float autoMoveTimeoutSeconds = 3.0f;

    // Events
    public event Action<string> OnReady;
    public event Action<string, List<string>> OnJoined;
    public event Action OnLeave;
    public event Action<Dictionary<string, JToken>> OnMoves;
    public event Action<string, string> OnMessage;

    // State
    public string Id { get; private set; }
    public string HostId { get; private set; }
    private bool connectedHost = false;
    private List<string> guests = new List<string>();

    private WebSocket ws;

    private float autoMoveTimer = 0f;
    private bool isAutoMoveTimerActive = false;

    private void Update()
    {
        if (ws != null)
        {
#if !UNITY_WEBGL || UNITY_EDITOR
            // En plataformas nativas o el Editor, despacha los mensajes al hilo principal
            ws.DispatchMessageQueue();
#endif
        }

        // Handle auto-move timer
        if (isAutoMoveTimerActive)
        {
            autoMoveTimer -= Time.deltaTime;
            if (autoMoveTimer <= 0)
            {
                isAutoMoveTimerActive = false;
                Move(null); // Send empty move to stay alive
            }
        }
    }

    private async void OnApplicationQuit()
    {
        if (ws != null)
        {
            await ws.Close();
        }
    }

    public async void Connect()
    {
        ws = new WebSocket(wsUrl);

        ws.OnOpen += () =>
        {
            // Request ID
            SendMessage(new { action = "get_id" });
        };

        ws.OnError += (errMsg) => Debug.LogError($"WebSocket Error: {errMsg}");
        ws.OnClose += (closeCode) => Debug.Log("WebSocket Closed");

        ws.OnMessage += (bytes) =>
        {
            string json = Encoding.UTF8.GetString(bytes);
            JObject data = JObject.Parse(json);
            string type = data["type"]?.ToString();

            if (type == "your_id")
            {
                Id = data["id"].ToString();
                OnReady?.Invoke(Id);
            }
            else if (type == "confirm_join")
            {
                connectedHost = true;
                HostId = data["hostId"].ToString();
                guests = data["guests"].ToObject<List<string>>();
                
                OnJoined?.Invoke(HostId, guests);
                StartAutoMoveTimer();
            }
            else if (type == "guest_joined")
            {
                string newGuestId = data["newGuestId"].ToString();
                guests.Add(newGuestId);
                Debug.Log($"Guest {newGuestId} joined the room.");
            }
            else if (type == "guest_left")
            {
                string leftGuestId = data["leftGuestId"].ToString();
                guests.Remove(leftGuestId);
                Debug.Log($"Guest {leftGuestId} left the room.");
            }
            else if (type == "turn_moves")
            {
                var moves = data["moves"].ToObject<Dictionary<string, JToken>>();
                OnMoves?.Invoke(moves);
                StartAutoMoveTimer();
            }
            else if (type == "kicked")
            {
                connectedHost = false;
                HostId = null;
                guests.Clear();
                isAutoMoveTimerActive = false;
                OnLeave?.Invoke();
            }
            else if (type == "direct_message")
            {
                string msg = data["message"].ToString();
                string senderId = data["senderId"].ToString();
                OnMessage?.Invoke(msg, senderId);
            }
        };

        await ws.Connect();
    }

    public async void Disconnect()
    {
        if (ws != null && ws.State == WebSocketState.Open)
        {
            await ws.Close();
        }
    }

    public void Join(string hostId)
    {
        if (string.IsNullOrEmpty(Id))
        {
            Debug.LogError("Not ready. Wait for OnReady callback.");
            return;
        }

        HostId = hostId;
        SendMessage(new
        {
            action = "route",
            targets = new[] { HostId },
            payload = new { type = "join", guestId = Id }
        });
    }

    public void Leave()
    {
        if (!string.IsNullOrEmpty(HostId) && connectedHost)
        {
            connectedHost = false;
            isAutoMoveTimerActive = false;
            OnLeave?.Invoke();
        }
    }

    public void Move(object moveData)
    {
        if (string.IsNullOrEmpty(HostId) || !connectedHost)
        {
            Debug.LogError("Not connected to a host.");
            return;
        }

        isAutoMoveTimerActive = false; // Cancel auto-move

        SendMessage(new
        {
            action = "route",
            targets = new[] { HostId },
            payload = new { type = "move", guestId = Id, moveData = moveData }
        });
    }

    public List<string> GetPlayers()
    {
        if (!connectedHost) return new List<string>();
        var players = new List<string> { HostId };
        players.AddRange(guests);
        return players;
    }

    public void SendDirectMessage(object message, List<string> targets = null)
    {
        if (!connectedHost) return;

        if (targets == null || targets.Count == 0)
        {
            targets = new List<string> { HostId };
            targets.AddRange(guests.Where(g => g != Id));
        }

        if (targets.Count > 0)
        {
            SendMessage(new
            {
                action = "route",
                targets = targets,
                payload = new
                {
                    type = "direct_message",
                    senderId = Id,
                    message = message
                }
            });
        }
    }

    private void StartAutoMoveTimer()
    {
        autoMoveTimer = autoMoveTimeoutSeconds;
        isAutoMoveTimerActive = true;
    }

    private async void SendMessage(object payload)
    {
        if (ws == null || ws.State != WebSocketState.Open) return;

        string json = JsonConvert.SerializeObject(payload);
        await ws.SendText(json);
    }
}

File: client/GameGuest.js
export class GameGuest {
  constructor(config = {}) {
    this.wsUrl = config.wsUrl;
    
    // Callbacks
    this.onJoined = config.onJoined || (() => {}); // Fired when host confirms join
    this.onLeave = config.onLeave || (() => {});   // Fired when kicked or host leaves
    this.onMoves = config.onMoves || (() => {});   // Fired on turn resolution
    this.onReady = config.onReady || (() => {});   // Fired when guest gets its ID
    this.onMessage = config.onMessage || (() => {}); // Fired on direct messages

    this.id = null;
    this.hostId = null;
    this.ws = null;
    this.connectedHost = false;
    this.guests = []; // List of all guests in the room including us
    
    this.autoMoveTimeout = config.autoMoveTimeout || 3000;
    this.moveTimer = null;
  }

  connect() {
    this.ws = new WebSocket(this.wsUrl);

    this.ws.onopen = () => {
      // Request connection ID
      this.ws.send(JSON.stringify({ action: 'get_id' }));
    };

    this.ws.onmessage = (event) => {
      const data = JSON.parse(event.data);
      
      if (data.type === 'your_id') {
        this.id = data.id;
        this.onReady(this.id);
      } else if (data.type === 'confirm_join') {
        this.connectedHost = true;
        this.guests = data.guests;
        this.onJoined(data.hostId, data.guests);
        this.startAutoMoveTimer();
      } else if (data.type === 'guest_joined') {
        this.guests.push(data.newGuestId);
        // Another guest joined (optional handling)
        console.log(`Guest ${data.newGuestId} joined the room.`);
      } else if (data.type === 'guest_left') {
        this.guests = this.guests.filter(g => g !== data.leftGuestId);
        // Another guest left (optional handling)
        console.log(`Guest ${data.leftGuestId} left the room.`);
      } else if (data.type === 'turn_moves') {
        this.onMoves(data.moves);
        this.startAutoMoveTimer();
      } else if (data.type === 'kicked') {
        this.connectedHost = false;
        this.hostId = null;
        this.guests = [];
        if (this.moveTimer) clearTimeout(this.moveTimer);
        this.onLeave();
      } else if (data.type === 'direct_message') {
        this.onMessage(data.message, data.senderId);
      }
    };
  }

  startAutoMoveTimer() {
    if (this.moveTimer) clearTimeout(this.moveTimer);
    this.moveTimer = setTimeout(() => {
      // Send an empty move to stay alive
      this.move(null);
    }, this.autoMoveTimeout);
  }

  getPlayers() {
    if (!this.connectedHost) return [];
    return [this.hostId, ...this.guests];
  }

  send(message, targets) {
    if (!this.connectedHost) return;
    
    if (!targets) {
      // Send to host and all other guests
      targets = [this.hostId, ...this.guests.filter(g => g !== this.id)];
    }
    
    if (targets.length > 0) {
      this._sendMessage(targets, { type: 'direct_message', senderId: this.id, message: message });
    }
  }

  join(hostId) {
    if (!this.id) {
      console.error("Not ready. Wait for onReady callback.");
      return;
    }
    this.hostId = hostId;
    this._sendMessage(this.hostId, { type: 'join', guestId: this.id });
  }

  leave() {
    if (this.hostId && this.connectedHost) {
      this.connectedHost = false;
      if (this.moveTimer) clearTimeout(this.moveTimer);
      this.onLeave();
      // We don't necessarily send a leave message to the host since the host
      // removes guests that timeout, but we could add it as an optimization later.
    }
  }

  move(moveData) {
    if (!this.hostId || !this.connectedHost) {
      console.error("Not connected to a host.");
      return;
    }
    if (this.moveTimer) clearTimeout(this.moveTimer);
    this._sendMessage(this.hostId, { type: 'move', guestId: this.id, moveData });
  }

  _sendMessage(targets, payload) {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;
    
    this.ws.send(JSON.stringify({
      action: 'route',
      targets: targets,
      payload: payload
    }));
  }
}
